src/macos/icons.rs

```diff
@@ -0,0 +1,96 @@
+use anyhow::{bail, Result};
+use cocoa::base::id;
+use image::ImageEncoder;
+use objc::{class, msg_send, sel, sel_impl};
+use once_cell::sync::Lazy;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::path::Path;
+use std::path::PathBuf;
+use std::{collections::hash_map::Entry, sync::Mutex};
+use sysinfo::{PidExt, ProcessExt, ProcessRefreshKind, System, SystemExt};
+
+pub static ICON_CACHE: Lazy<Mutex<IconCache>> = Lazy::new(|| Mutex::new(IconCache::default()));
+
+#[derive(Default)]
+pub struct IconCache {
+    /// executable name -> icon hash
+    executables: HashMap<PathBuf, u64>,
+    /// icon hash -> png bytes
+    icons: HashMap<u64, Vec<u8>>,
+}
+
+impl IconCache {
+    pub fn get_png(&mut self, executable: PathBuf) -> Result<&Vec<u8>> {
+        match self.executables.entry(executable) {
+            Entry::Occupied(e) => {
+                // Guaranteed to exist because we never clear the cache.
+                Ok(self.icons.get(e.get()).unwrap())
+            }
+            Entry::Vacant(e) => {
+                let tif = unsafe { tif_data_for_executable(e.key())? };
+                let icon = tif_to_png(&tif)?;
```

Comment:
This makes no sense here yet. I think we want to 1) get the raw TIF bytes, and 2) hash them, 3) check if the PNG is in cache, or 4) convert it. We don't want to do the conversion everytime.